KAFKA-16339: [1/4 KStream#transform] Remove Deprecated "transformer" methods and classes

fonsdant opened the pull request:
Hi, @mjsax! I have removed the `transform` methods from the main and test classes. However, I have asked myself whether the approach I have used for `StandbyTaskEOSIntegrationTest` and `StreamThreadTest` is correct. Should I replace `transform` with `process` as suggested in `transform` Javadoc or remove these tests? Thanks!

fonsdant commented:
I have missed removing the references to `transform` on Javadocs too. I will push another commit for it.

fonsdant commented:
I have noticed that `KStreamTransformIntegrationTest.TestTransformer` could be removed as it has become unused.

fonsdant commented:
For now, I have removed the build with deduplicaiton topology test until we decide how to replace the `transform` method inside it.

fonsdant commented:
I have replaced `tranform` with `process` as suggested by docs. I have implemented Processor based on `StandbyTaskEOSMultiRebalanceIntegrationTest.buildWithUniqueIdAssignmentTopology`, but `shouldWipeOutStandbyStateDirectoryIfCheckpointIsMissing` has failed.

fonsdant commented:
I have managed to fix the test! :D

fonsdant commented:
I have removed the references to `transform` from docs. I am wondering if it would be better to replace `transform` with `process`, instead of removing it. I have started to think about it while removing the `flatTransformValues` (the third method to remove). @mjsax, could you help me with this decision?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -179,8 +178,7 @@ <KR> KStream<KR, V> selectKey(final KeyValueMapper<? super K, ? super V, ? exten
      * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K':V'>}.
-     * This is a stateless record-by-record operation (cf. {@link #transform(TransformerSupplier, String...)} for
-     * stateful record transformation).
+     * This is a stateless record-by-record operation.
```
Yes, it think we should link to `#process(...)` here, instead of removing (same elsewhere where it make sense).

Btw: feel free to skip updating JavaDocs of other deprecated methods which will be remove in follow up PRs of this ticket.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java:
```diff
@@ -43,7 +43,6 @@
  * @param <R> {@link KeyValue} return type (both key and value type can be set
  *            arbitrarily)
  * @see TransformerSupplier
- * @see KStream#transform(TransformerSupplier, String...)
```
We will also remove `Transformer` in a follow up PR completely, right (please verify...)? So no need to have any updates in this PR?

fonsdant replied:
Although we remove it, I think it helpful to update it to prevent `gradle javadoc` to fail. I am checking if the modifications are right this way. Is there any problem if I go this way? Or some suggestion?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/TransformerSupplier.java:
```diff
@@ -33,7 +33,6 @@
  * @param <R> {@link org.apache.kafka.streams.KeyValue KeyValue} return type (both key and value type can be set
  *            arbitrarily)
  * @see Transformer
- * @see KStream#transform(TransformerSupplier, String...)
```
As for `Transformer`

fonsdant replied:
Although we remove it, I think it helpful to update it to prevent `gradle javadoc` to fail. I am checking if the modifications are right this way. Is there any problem if I go this way? Or some suggestion?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java:
```diff
@@ -35,7 +35,6 @@
  * @see KStream#transformValues(ValueTransformerWithKeySupplier, String...)
  * @see Transformer
  * @see TransformerSupplier
- * @see KStream#transform(TransformerSupplier, String...)
```
This does make sense, as we don't remove `ValueTransformerSupplier` yet.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -1243,24 +1243,6 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
-                                              final String... stateStoreNames) {
-        Objects.requireNonNull(transformerSupplier, "transformerSupplier can't be null");
-        final String name = builder.newProcessorName(TRANSFORM_NAME);
-        return flatTransform(new TransformerSupplierAdapter<>(transformerSupplier), Named.as(name), stateStoreNames);
```
`TransformerSupplierAdapter` should be unused now, and I think we can remove it with this PR, too, including `TransformerSupplierAdapterTest`?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -329,26 +331,32 @@ private KafkaStreams buildWithDeduplicationTopology(final String stateDirPath) {
 
                         if (store.get(key) != null) {
-                            return null;
+                            store.delete(key);
```
Why do we insert a delete here?

fonsdant replied:
I have thought it would be the most similar replacement for return null. Since we are not returning null anymore, we are not having it in store.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -360,11 +368,9 @@ public KeyValue<Integer, Integer> transform(final Integer key, final Integer val
                         }
 
-                        return KeyValue.pair(key, value);
+                        store.put(key, value);
```
We do we add a put here?

fonsdant replied:
Similar to the return null replacement with delete. Since we would return a pair with key and value, now I put it in store.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -2334,85 +2332,6 @@ public void process(final Record<Object, Object> record) {}
-    @ParameterizedTest
-    @MethodSource("data")        
-    @SuppressWarnings("deprecation")
-    public void shouldPunctuateWithTimestampPreservedInProcessorContext(final boolean stateUpdaterEnabled, final boolean processingThreadsEnabled) {
```
I think we need to keep this test, but rewrite using `#process()`.

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -276,61 +274,6 @@ class TopologyTest {
 
-  @nowarn
-  @Test
-  def shouldBuildIdenticalTopologyInJavaNScalaTransform(): Unit = {
```
Wondering if we have a test gap, as we don't have a similar test for `process()` -- should we add a new test for `#process()` like this one (or rewrite this test for `#process()`)?

fonsdant replied:
I have rewritten this test with `process`, but have missed to rewrite the test name. I will fix it.

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -222,42 +221,6 @@ class KStreamTest extends TestDriver {
 
-  @nowarn
-  @Test
-  def testTransformCorrectlyRecords(): Unit = {
```
Same question as above. Seems we are lacking a similar test for `#process()` ?

fonsdant replied:
Hm... I have replaced Transformer with ProcessorSupplier and transform with process. Have I missed something else?

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -447,28 +410,4 @@ class KStreamTest extends TestDriver {
-  @nowarn
-  @Test
-  def testSettingNameOnTransform(): Unit = {
-    class TestTransformer extends Transformer[String, String, KeyValue[String, String]] {
```
Same

mjsax reviewed (commented):
Thanks for the PR. Made a pass.

Btw: is there anything in `docs/streams/...` that we need to update/remove, too (we can also do this as a single docs-only PR for the whole ticket, ie, all `transformXxx()` method)

fonsdant commented:
I will push some commits, one for each file from which I have removed the transform link, replacing it with process.

fonsdant committed:
remove transform from kstream

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove transform from kstream impl

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove transform from kstream scala

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
I have rebased my branch on trunk and this caused GitHub to close the PR. I am implementing the last review points again.

fonsdant committed:
remove kstream transform integration test

fonsdant committed:
remove kstream transform test

fonsdant committed:
remove transform from streams builder test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove transform from kstream impl test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove transformer supplier adapter

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove transformer supplier adapter test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
This time, I have focused on removing the methods first and doing refactorings. After this, I will proceed with Javadoc updates.

I have made the removing part. Now, I will do the refactoring part.

mjsax reviewed (approved):
Thanks for updating the PR. LGTM.

Just triggered CI. Can merge after it finished and passed.

mjsax commented:
@fonsdant -- seems there is some compilation error:
```
StandbyTaskEOSIntegrationTest.java:332: error: cannot find symbol

            .transform(
> Task :streams:compileTestJava
            ^
  symbol:   method transform(()->new or[...]{ } },String)
  location: interface KStream<Integer,Integer>
  ```
  
Seem we need to rewrite this test to use the new `process()` instead of `transform()`

fonsdant commented:
Thanks, @mjsax, still need to refactor it. Meanwhile, do you prefer that I convert this PR to draft? Or should I keep it as it is?

mjsax commented:
We can keep this PR as-is -- you can also to the refactoring as part of this PR if you want.

mjsax commented:
@fonsdant Any update on this PR? -- We are approaching AK 4.0 release deadlines, and there is 3 more PRs for this ticket...

fonsdant commented:
@mjsax, sorry for the delay, I am still refactoring the tests, replacing transform with process.

Given the coming deadline, I agree it would be adequate someone more experienced to resume it.

fonsdant commented:
I will push the work I done this passed week.

mjsax commented:
> Given the coming deadline, I agree it would be adequate someone more experienced to resume it.

As long as you have enough time to keep working, I think we are a good. Of course, if you are blocked, let us know.

fonsdant committed:
refator stand by task eos integration test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
refator stream thread test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
refactor scala topology test

fonsdant committed:
refactor scala kstream test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
I have managed to refactor the tests! 😄 

Now, I will just update the Javadocs and Scaladocs again

fonsdant committed:
adjust topology test for spotless scala

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
adjust kstream test for spotless scala

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of transformer

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of transformer supplier

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of value transformer supplier

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of value transformer with key supplier

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of connected store provider

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
update javadoc of kstream

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

mjsax commented:
@fonsdant There is merge conflicts -- can you rebase to latest `trunk`?

fonsdant commented:
@mjsax, rebased!

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -329,26 +330,32 @@ private KafkaStreams buildWithDeduplicationTopology(final String stateDirPath) {
-                    public KeyValue<Integer, Integer> transform(final Integer key, final Integer value) {
+                    public void process(final Record<Integer, Integer> record) {
+                        final var key = record.key();
+                        final var value = record.value();
```
Why would we use `var` here? We know the type and thus should use `int`? `var` should only be used for lambdas in which we might not know the type.

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -329,26 +330,32 @@ private KafkaStreams buildWithDeduplicationTopology(final String stateDirPath) {
                             // we only forward so we can verify the skipping by reading the output topic
                             // the goal is skipping is to not modify the state store
-                            return KeyValue.pair(key, value);
+                            store.put(key, value);
```
Why do we add  this line? It was not there in the original code.

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -329,26 +330,32 @@ private KafkaStreams buildWithDeduplicationTopology(final String stateDirPath) {
 
                         if (store.get(key) != null) {
-                            return null;
+                            store.delete(key);
```
Why do we add this line? Seem we just want to `return;` here to exit early?

mjsax commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java:
```diff
@@ -360,11 +367,9 @@ public KeyValue<Integer, Integer> transform(final Integer key, final Integer val
                         }
 
-                        return KeyValue.pair(key, value);
+                        store.put(key, value);
```
Why do we add this line?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -2289,34 +2289,29 @@ public void process(final Record<Object, Object> record) {}
+        final ProcessorSupplier<Object, Object, Object, Object> punctuateProcessor = () -> new Processor<>() {
+            @Override
+            public void init(final ProcessorContext<Object, Object> context) {
+                final var duration = Duration.ofMillis(100L);
```
```suggestion
                final Duration duration = Duration.ofMillis(100L);
```

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -448,27 +478,27 @@ class KStreamTest extends TestDriver {
+//
+//    val transformNode = builder.build().describe().subtopologies().asScala.head.nodes().asScala.toList(1)
+//    assertEquals("my-name", transformNode.name())
+//  }
```
Should we rewrite this to `testSettingNameOnProcess`?

fonsdant replied:
Oh, sorry! I have missed it up. Of course, I will rewrite it!

fonsdant committed:
refactor standby task eos integration test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
refactor stream thread test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
refactor scala kstream test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
@mjsax, I have removed the adding lines in buildTopologyWithDeduplication in StandbyTaskEOSIntegrationTest to match the previous implementation with transform (but still with process). Also, removed and rewritten the commented test. Finally, removed var usage.

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -478,27 +479,51 @@ class KStreamTest extends TestDriver {
+  @Test
+  def testSettingNameOnTransform(): Unit = {
+    val processorSupplier: ProcessorSupplier[String, String, String, String] =
+      new api.ProcessorSupplier[String, String, String, String] {
```
Why do we need an explicit `ProcessorSupplier` for this test? And we use `process(() => new TestProcessor)` similar to the old code?

fonsdant replied:
Oh, it is true... I have made this way, because IntelliJ highlight the "TestProcessor" as an error --- probably conflicting with the deprecated `process`. See:

![image](https://github.com/user-attachments/assets/582b86e0-232d-4c68-ab18-0faebb3060ae)

But I can actually run the test this way:

![image](https://github.com/user-attachments/assets/adf5a3df-8f3d-4fbe-90c7-05a1514f743c)

mjsax replied:
Not sure why IntelliJ shows an error... given that `TestProcessor` is `api.Processor` I would expect IntelliJ to figure out that the lambda is `api.ProcessorSupplier` and resolve to the new `process(...)` overload... In any case, the build did pass, so it seem ok. Also, the old `process()` will be removed soon, too: https://github.com/apache/kafka/pull/17190

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -478,27 +479,51 @@ class KStreamTest extends TestDriver {
+  def testSettingNameOnTransform(): Unit = {
+    val processorSupplier: ProcessorSupplier[String, String, String, String] =
+      new api.ProcessorSupplier[String, String, String, String] {
+        private val storeName = "store-name"
```
Why do we add a store?

mjsax commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala:
```diff
@@ -478,27 +479,51 @@ class KStreamTest extends TestDriver {
+              val processedKey = s"$key-processed"
+              val processedValue = s"$value-processed"
+              store.put(processedKey, processedValue)
+              context.forward(new api.Record(processedKey, processedValue, record.timestamp()))
```
What the `Processor` does is actually pretty meaningless for the test, so I think the while implementation could actually be empty? For this case, we don't even need `init()` method, as we don't need access to the `context` either.

mjsax reviewed (commented):
Thanks for the quick turn around. I think the Scala test can be simplified significantly. In  the end, we only verify the name of he processor node in the `Topology`.

fonsdant committed:
refactor scala kstream test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
Thanks, @mjsax! I have simplified the test and replaced the processorSupplier with a lambda. Should I replace it for for other methods too? And how about IntelliJ reporting it as problem?

mjsax commented:
Merged to `trunk`. Thanks for the PR! One down. Three to go. :)  Looking forward to your update of https://github.com/apache/kafka/pull/17245

fonsdant commented:
Uhuu! Here we go! I need to rebase the second one on this first now. Thanks, @mjsax!
